src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java

```diff
@@ -1337,159 +1369,402 @@ public void setMirror()
+    {
+        return getOpenRequests(citizen).stream().anyMatch(selectionPredicate);
+    }
+
+    public <R> boolean hasWorkerOpenRequestsOfType(@NotNull final CitizenData citizenData, final TypeToken<R> requestType)
+    {
+        return !getOpenRequestsOfType(citizenData, requestType).isEmpty();
+    }
+
+    @SuppressWarnings(Suppression.GENERIC_WILDCARD)
+    public <R> ImmutableList<IRequest<? extends R>> getOpenRequestsOfType(
+                                                                           @NotNull final CitizenData citizenData,
+                                                                           final TypeToken<R> requestType)
+    {
+        return ImmutableList.copyOf(getOpenRequests(citizenData).stream()
+                                      .filter(request -> {
+                                          Set<TypeToken> requestTypes = ReflectionUtils.getSuperClasses(request.getRequestType());
+                                          return requestTypes.contains(requestType);
+                                      })
+                                      .map(request -> (IRequest<? extends R>) request)
+                                      .iterator());
+    }
+
+    public boolean hasCitizenCompletedRequests(@NotNull final CitizenData data)
+    {
+        return !getCompletedRequests(data).isEmpty();
+    }
+
+    public ImmutableList<IRequest> getCompletedRequests(@NotNull final CitizenData data)
+    {
+        if (!citizensByCompletedRequests.containsKey(data.getId()))
         {
-            return location;
+            return ImmutableList.of();
         }
 
-        /**
-         * Gets the ColonyView that this building belongs to.
-         *
-         * @return ColonyView, client side interpretations of Colony.
-         */
-        public ColonyView getColony()
+        return ImmutableList.copyOf(citizensByCompletedRequests.get(data.getId()).stream()
+                                      .map(getColony().getRequestManager()::getRequestForToken).filter(Objects::nonNull).iterator());
+    }
+
+    public <R> ImmutableList<IRequest<? extends R>> getCompletedRequestsOfType(@NotNull final CitizenData citizenData, final TypeToken<R> requestType)
+    {
+        return ImmutableList.copyOf(getCompletedRequests(citizenData).stream()
+                                      .filter(request -> {
+                                          Set<TypeToken> requestTypes = ReflectionUtils.getSuperClasses(request.getRequestType());
+                                          return requestTypes.contains(requestType);
+                                      })
+                                      .map(request -> (IRequest<? extends R>) request)
+                                      .iterator());
+    }
+
+    public <R> ImmutableList<IRequest<? extends R>> getCompletedRequestsOfTypeFiltered(
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Remove usage of generic wildcard type. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1452)